add chalk and yargs

arranf opened the pull request:
This PR adds the following dependencies:

* chalk for formatting (v4 due to v5 requiring ESM support)
* yargs for argument parsing

It also adds a `.nvmrc` file for standardising on a version of nvm, adds a script for specifically running the cli without a global `ts-node`.

I've set the commands to be backwards compatible but also to make `create-plugin` its own command.

Here's a dodgy screenshot of how the chalk formatted output looks
![image](https://user-images.githubusercontent.com/8703222/159094095-63d64adf-95b3-4ff7-a645-819e16f0be77.png)

arranf committed:
add chalk and yargs

hpx7 commented:
This looks amazing, thanks @arranf!

A couple of small things:
1) I just added a `.prettierrc` to the develop branch, can you update and use that? that way we can get rid of the formatting changes in the PR
2) Can you open a separate PR for the `.nvmrc` file and another one for the script change? That way we can keep this PR focused on just the chalk/yargs stuff

arranf committed:
Merge branch 'develop' of github.com:arranf/hathora into feature/cli-framework

arranf committed:
name CLI hathora

arranf committed:
reformat cli.ts and remove .nvmrc

arranf committed:
fix readme

arranf commented:
> This looks amazing, thanks @arranf!
> 
> A couple of small things:
> 
>     1. I just added a `.prettierrc` to the develop branch, can you update and use that? that way we can get rid of the formatting changes in the PR
> 
>     2. Can you open a separate PR for the `.nvmrc` file and another one for the script change? That way we can keep this PR focused on just the chalk/yargs stuff

1. Done!
2. I've deleted the `.nvmrc` changes and I'll open up a new PR for that, I'd like to keep the addition of the the new `cli" script part of this PR though - as I'd consider it part of the scope of the PR which is improving the CLI

hpx7 commented on .gitignore:
```diff
@@ -1,3 +1,4 @@
 /*.js
 node_modules/
 .DS_Store
+hathora.yml
```
what was the reason for adding this?

arranf replied:
I ended up with a `hathora.yml` in my repo through the process of testing and almost checked it in - I figured that this wouldn't hurt and might prevent a mistake in the future!

hpx7 replied:
i think this could cause issues with the example apps in this repo since they have a `hathora.yml` file which is supposed to be checked in

arranf replied:
If I scoped it to the project root would that be acceptable?

hpx7 replied:
i would prefer if we just got rid of it

hpx7 commented on README.md:
````diff
@@ -66,14 +66,8 @@ To contribute to Hathora, first clone the repo:
 ```
 
-Make sure you have `ts-node` installed globally:
+You can now your local hathora cli as follows:
````
typo
```suggestion
You can now run your local hathora cli as follows:
```

hpx7 commented on cli.ts:
```diff
@@ -106,7 +106,10 @@ function build() {
       buildClient({
         root: join(clientDir, dir),
-        build: { outDir: join(rootDir, "dist", "client", dir), target: ["esnext"] },
+        build: {
+          outDir: join(rootDir, "dist", "client", dir),
+          target: ["esnext"],
+        },
```
can revert this formatting change

arranf replied:
Done.

hpx7 commented on cli.ts:
```diff
@@ -127,51 +130,140 @@ const rootDir = getProjectRoot(process.cwd());
+      install();
+    },
+  })
+  .command({
+    // For backwards compatibility
+    command: "create-plugin-native <type>",
+    describe: "Creates a native plugin",
+    handler: (argv) => {
+      generate(rootDir, "templates/plugin/native", {
+        val: argv.type as string,
+      });
+    },
+  })
+  .command({
+    // For backwards compatibility
+    command: "create-plugin-react <type>",
+    describe: "Creates a react plugin",
+    handler: (argv) => {
+      generate(rootDir, "templates/plugin/react", { val: argv.type as string });
+    },
+  })
+  .command({
+    // For backwards compatibility
+    command: "create-plugin-view <type>",
+    aliases: ["create-plugin-lit"],
+    describe: "Creates a lit plugin",
+    handler: (argv) => {
+      generate(rootDir, "templates/plugin/lit", { val: argv.type as string });
+    },
+  })
```
we don't need to maintain compatibility for these commands at this stage, just need to update the docs with the new version of the command

arranf replied:
I've kept them in, updated the docs, and marked them as deprecated. I'd suggest we remove them in a future version.

We should add some validation or error handling around what happens if someone inputs something other than `react`, `lit`, and `native`. Right now they'll get a missing directory error I think.

Any thoughts on what kind of error handling?

hpx7 replied:
is it possible to tell yargs what arguments to expect? would be nice if it was included in the help text

hpx7 commented on package.json:
```diff
@@ -14,6 +14,7 @@
   ],
   "dependencies": {
     "@node-loader/core": "2.0.0",
+    "chalk": "^4.1.2",
```
i prefer to use exact version syntax for all the hathora deps (been burned too many times by package breaks)

kasbah replied:
That's what the package-lock.json is for.

hpx7 replied:
package-lock is good for local dependencies but it doesn't get published to npm, so it doesn't apply to folks who install hathora

hpx7 replied:
still need to switch `chalk` and `@types/yargs` to exact version syntax

hpx7 commented:
@arranf should we also add a version command?
```js
.version(function() { return require('../package').version; })
```

arranf committed:
Update README.md

Co-authored-by: Harsh Pandey <hpandey793@gmail.com>

arranf committed:
mark old create-plugin as deprecated

arranf committed:
Merge branch 'feature/cli-framework' of github.com:arranf/hathora into feature/cli-framework

arranf commented:
> @arranf should we also add a version command?
> 
> ```js
> .version(function() { return require('../package').version; })
> ```

Added.

hpx7 commented on package.json:
```diff
@@ -32,6 +34,8 @@
     "@types/shelljs": "0.8.11",
     "@types/uuid": "8.3.4",
+    "@types/yargs": "^17.0.10",
+    "prettier": "^2.6.0",
```
i think we can remove `prettier` for now add it later when we actually integrate it into CI etc

arranf replied:
I've added it because my workflow involves trying to add as few global packages to my namespace as possible (to avoid conflicting versions in my global namespace). Adding `prettier` here means I can invoke prettier from dev dependencies vs having to install it globally.

hpx7 replied:
oh I see, I've just been using vscode which has prettier built in

hpx7 committed:
Merge branch 'develop' into feature/cli-framework
